license_manager/service/license.go

```diff
@@ -0,0 +1,141 @@
+package service
+
+import (
+	"strings"
+	"time"
+
+	lmpb "github.com/enfabrica/enkit/license_manager/proto"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// license manages allocations and queued invocations for a single license
+// type.
+type license struct {
+	totalAvailable int                    // Constant total number of licenses available for invocations.
```

Comment:
These are the license counts:

1. Xilinx: 3 Vivado tool licenses
2. Cadence: 6 Xcelium tool licenses, 5 Jasper tool licenses
3. Synopsys: 1 Fusion Compiler tool license, 2 VC static tool licenses, 2 Spyglass tool licenses, 1 VCS tool license, 1 Verdi tool license

Different EDA tool vendors contain a different number of license counts per tool.

Reply:
Sorry for the confusion. Let me clarify my comment.

There are two bits of information:
1) How many licenses we have available in total, and which kind.
2) How many licenses are allocated.

I still stand that by my original statement that allocation/deallocation should be out of band. Monitoring will tell us if there are discrepancies, and we'll react to that.

Several reasons for this, but the main one is that through a single build, some builds allocate/release the license in the middle of the build step. But if a license is not available in the middle of the step, the step can crash/fail. 

So, ... let's say you query the license server in the middle of such a build, it will likely say "this client is using 0 licenses", but in practice, 1 license MUST be reserved for the whole duration of the step. Our license server does the reservation.

That said... the TOTAL number of license AVAILABLE, the licenses we paid for, should be a pretty constant number, and both our "reservation system" and "license manager" need to somewhat agree on that number.

So far, the # of TOTAL licenses was hardcoded in the binary. Having a config file (or a built in config file, like we do for other binaries), is a good plan.

The annoying part is that whenever licenses expire, or we need to buy new ones, we need to update two places: the real license manager, and our own. Which is error prone.

So my thought was really to get the TOTAL # of licenses (at startup, or periodically) from the license server, so we don't have to maintain this second configuration file, and we can no longer make mistakes here.

Reply:
Thanks for the clarification - that makes sense.

I think overall that would be a good quality-of-life improvement - we can add that as that becomes the biggest issue? I think right now it's still an open question whether the current approach can result in something stable given significant build load as-is.

Reply:
sure, sgtm, yes. - future PR makes sense.